File: go/poker.go
package main

import (
    "bufio"
    "fmt"
    "math/rand"
    "os"
    "sort"
    "strings"
    "time"
)

type Card struct {
    face string
    suit string
}

func (c Card) toString() string {
    return c.face + c.suit
}

func main() {
    rand.Seed(time.Now().UnixNano())

    faces := []string{"2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"}
    suits := []string{"D", "C", "H", "S"}

    var deck []Card
    for _, face := range faces {
        for _, suit := range suits {
            deck = append(deck, Card{face, suit})
        }
    }

    if len(os.Args) == 1 {
        fmt.Println("*** P O K E R H A N D A N A L Y Z E R ***")
        fmt.Println("*** USING RANDOMIZED DECK OF CARDS ***")

        // Shuffle the deck
        fisherYatesShuffle(deck)

        fmt.Println("*** Shuffled 52 card deck:")
        printDeck(deck)

        fmt.Println("*** Here are the six hands...")
        hands := dealHands(deck, 6, 5)
        for _, hand := range hands {
            printHand(hand)
        }

        fmt.Println("*** Here is what remains in the deck...")
        printDeck(deck)

        fmt.Println("--- WINNING HAND ORDER ---")
        sortedHands := sortHands(hands)
        for _, hand := range sortedHands {
            printHand(hand)
            fmt.Printf(" - %s\n", evaluateHand(hand))
        }
    } else {
        filename := os.Args[1]
        fmt.Println("*** P O K E R H A N D A N A L Y Z E R ***")
        fmt.Println("*** USING TEST DECK ***")
        fmt.Printf("*** File: %s\n", filename)

        file, err := os.Open(filename)
        if err != nil {
            fmt.Printf("Could not open file '%s' %v", filename, err)
            return
        }
        defer file.Close()

        var hands [][]Card
        scanner := bufio.NewScanner(file)
        for scanner.Scan() {
            line := scanner.Text()
            fmt.Println(line)
            cards := strings.Split(line, ", ")
            var hand []Card
            for _, card := range cards {
                card = strings.TrimSpace(card)
                face := card[:len(card)-1]
                suit := card[len(card)-1:]
                hand = append(hand, Card{face, suit})
            }
            hands = append(hands, hand)
        }

        fmt.Println("*** Here are the six hands...")
        for _, hand := range hands {
            printHand(hand)
        }

        fmt.Println("--- WINNING HAND ORDER ---")
        sortedHands := sortHands(hands)
        for _, hand := range sortedHands {
            printHand(hand)
            fmt.Printf(" - %s\n", evaluateHand(hand))
        }
    }
}

func fisherYatesShuffle(deck []Card) {
    for i := len(deck) - 1; i > 0; i-- {
        j := rand.Intn(i + 1)
        deck[i], deck[j] = deck[j], deck[i]
    }
}

func dealHands(deck []Card, numHands, cardsPerHand int) [][]Card {
    var hands [][]Card
    for i := 0; i < numHands; i++ {
        hand := deck[:cardsPerHand]
        deck = deck[cardsPerHand:]
        hands = append(hands, hand)
    }
    return hands
}

func printDeck(deck []Card) {
    for i, card := range deck {
        fmt.Print(card.toString() + " ")
        if (i+1)%13 == 0 {
            fmt.Println()
        }
    }
    fmt.Println()
}

func printHand(hand []Card) {
    for _, card := range hand {
        fmt.Print(card.toString() + " ")
    }
    fmt.Println()
}

func evaluateHand(hand []Card) string {
    faces := make(map[string]int)
    suits := make(map[string]int)
    for _, card := range hand {
        faces[card.face]++
        suits[card.suit]++
    }

    uniqueFaces := make([]string, 0, len(faces))
    for face := range faces {
        uniqueFaces = append(uniqueFaces, face)
    }
    uniqueSuits := make([]string, 0, len(suits))
    for suit := range suits {
        uniqueSuits = append(uniqueSuits, suit)
    }

    isFlush := len(uniqueSuits) == 1
    isStraight := false
    if len(uniqueFaces) == 5 {
        sort.Slice(uniqueFaces, func(i, j int) bool {
            return faces[uniqueFaces[i]] < faces[uniqueFaces[j]]
        })
        isStraight = true
        for i := 1; i < 5; i++ {
            if faces[uniqueFaces[i]] != faces[uniqueFaces[i-1]]+1 {
                isStraight = false
                break
            }
        }
        if !isStraight && uniqueFaces[0] == "A" && uniqueFaces[1] == "2" && uniqueFaces[2] == "3" && uniqueFaces[3] == "4" && uniqueFaces[4] == "5" {
            isStraight = true
        }
    }

    if isFlush && isStraight {
        sort.Slice(uniqueFaces, func(i, j int) bool {
            return faces[uniqueFaces[i]] < faces[uniqueFaces[j]]
        })
        if uniqueFaces[0] == "10" && uniqueFaces[4] == "A" {
            return "Royal Straight Flush"
        } else {
            return "Straight Flush"
        }
    } else if len(uniqueFaces) == 2 {
        sort.Slice(uniqueFaces, func(i, j int) bool {
            return faces[uniqueFaces[i]] > faces[uniqueFaces[j]]
        })
        if faces[uniqueFaces[0]] == 4 {
            return "Four of a Kind"
        } else {
            return "Full House"
        }
    } else if isFlush {
        return "Flush"
    } else if isStraight {
        return "Straight"
    } else if len(uniqueFaces) == 3 {
        sort.Slice(uniqueFaces, func(i, j int) bool {
            return faces[uniqueFaces[i]] > faces[uniqueFaces[j]]
        })
        if faces[uniqueFaces[0]] == 3 {
            return "Three of a Kind"
        } else {
            return "Two Pair"
        }
    } else if len(uniqueFaces) == 4 {
        return "Pair"
    } else {
        return "High Card"
    }
}

func sortHands(hands [][]Card) [][]Card {
    sort.Slice(hands, func(i, j int) bool {
        rankA := getRank(evaluateHand(hands[i]))
        rankB := getRank(evaluateHand(hands[j]))

        if rankA == rankB {
            sortedA := make([]Card, len(hands[i]))
            copy(sortedA, hands[i])
            sortedB := make([]Card, len(hands[j]))
            copy(sortedB, hands[j])

            sort.Slice(sortedA, func(i, j int) bool {
                return sortedA[i].face > sortedA[j].face
            })
            sort.Slice(sortedB, func(i, j int) bool {
                return sortedB[i].face > sortedB[j].face
            })

            for k := 0; k < 5; k++ {
                if sortedA[k].face != sortedB[k].face {
                    return sortedB[k].face > sortedA[k].face
                } else if sortedA[k].suit != sortedB[k].suit {
                    return sortedB[k].suit > sortedA[k].suit
                }
            }

            return false
        } else {
            return rankB > rankA
        }
    })

    return hands
}

func getRank(handType string) int {
    ranks := map[string]int{
        "Royal Straight Flush": 10,
        "Straight Flush":       9,
        "Four of a Kind":       8,
        "Full House":           7,
        "Flush":                6,
        "Straight":             5,
        "Three of a Kind":      4,
        "Two Pair":             3,
        "Pair":                 2,
        "High Card":            1,
    }

    return ranks[handType]
}

File: perl/poker.pl
#!/usr/bin/perl

use strict;
use warnings;

# Card class
package Card;

sub new {
    my ($class, $face, $suit) = @_;
    my $self = {
        face => $face,
        suit => $suit,
    };
    bless $self, $class;
    return $self;
}

sub getFace {
    my ($self) = @_;
    return $self->{face};
}

sub getSuit {
    my ($self) = @_;
    return $self->{suit};
}

sub toString {
    my ($self) = @_;
    return $self->{face} . $self->{suit};
}

# Main program
package main;

my @faces = ('2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A');
my @suits = ('D', 'C', 'H', 'S');

my @deck = ();
foreach my $face (@faces) {
    foreach my $suit (@suits) {
        push @deck, Card->new($face, $suit);
    }
}

if (@ARGV == 0) {
    print "*** P O K E R H A N D A N A L Y Z E R ***\n";
    print "*** USING RANDOMIZED DECK OF CARDS ***\n";
    
    # Shuffle the deck
    fisher_yates_shuffle(\@deck);
    
    print "*** Shuffled 52 card deck:\n";
    printDeck(@deck);
    
    print "*** Here are the six hands...\n";
    my @hands = dealHands(\@deck, 6, 5);
    foreach my $hand (@hands) {
        printHand(@$hand);
    }
    
    print "*** Here is what remains in the deck...\n";
    printDeck(@deck);
    
    print "--- WINNING HAND ORDER ---\n";
    my @sortedHands = sortHands(@hands);
    foreach my $hand (@sortedHands) {
        printHand(@$hand);
        print " - " . evaluateHand(@$hand) . "\n";
    }
}
else {
    my $filename = $ARGV[0];
    print "*** P O K E R H A N D A N A L Y Z E R ***\n";
    print "*** USING TEST DECK ***\n";
    print "*** File: $filename\n";
    
    open(my $fh, '<', $filename) or die "Could not open file '$filename' $!";
    my @hands = ();
    while (my $line = <$fh>) {
        chomp $line;
        print "$line\n";
        my @cards = split(/,\s*/, $line);
        my @hand = ();
        foreach my $card (@cards) {
            $card =~ s/\s+//g;
            my $face = substr($card, 0, length($card) - 1);
            my $suit = substr($card, -1);
            push @hand, Card->new($face, $suit);
        }
        push @hands, \@hand;
    }
    close $fh;
    
    print "*** Here are the six hands...\n";
    foreach my $hand (@hands) {
        printHand(@$hand);
    }
    
    print "--- WINNING HAND ORDER ---\n";
    my @sortedHands = sortHands(@hands);
    foreach my $hand (@sortedHands) {
        printHand(@$hand);
        print " - " . evaluateHand(@$hand) . "\n";
    }
}

sub fisher_yates_shuffle {
    my ($array) = @_;
    for (my $i = @$array - 1; $i > 0; $i--) {
        my $j = int(rand($i + 1));
        @$array[$i, $j] = @$array[$j, $i];
    }
}

sub dealHands {
    my ($deck, $numHands, $cardsPerHand) = @_;
    my @hands = ();
    for (my $i = 0; $i < $numHands; $i++) {
        my @hand = ();
        for (my $j = 0; $j < $cardsPerHand; $j++) {
            push @hand, shift @$deck;
        }
        push @hands, \@hand;
    }
    return @hands;
}

sub printDeck {
    my (@deck) = @_;
    my $count = 0;
    foreach my $card (@deck) {
        print $card->toString() . " ";
        $count++;
        if ($count % 13 == 0) {
            print "\n";
        }
    }
    print "\n";
}

sub printHand {
    my (@hand) = @_;
    foreach my $card (@hand) {
        print $card->toString() . " ";
    }
    print "\n";
}

sub evaluateHand {
    my (@hand) = @_;
    
    my %faces = ();
    my %suits = ();
    foreach my $card (@hand) {
        $faces{$card->getFace()}++;
        $suits{$card->getSuit()}++;
    }
    
    my @uniqueFaces = keys %faces;
    my @uniqueSuits = keys %suits;
    
    my $isFlush = scalar @uniqueSuits == 1;
    my $isStraight = 0;
    if (scalar @uniqueFaces == 5) {
        my @sortedFaces = sort { $faces{$a} <=> $faces{$b} } @uniqueFaces;
        $isStraight = 1;
        for (my $i = 1; $i < 5; $i++) {
            if ($faces{$sortedFaces[$i]} != $faces{$sortedFaces[$i-1]} + 1) {
                $isStraight = 0;
                last;
            }
        }
        if (!$isStraight && $sortedFaces[0] eq 'A' && $sortedFaces[1] eq '2' && $sortedFaces[2] eq '3' && $sortedFaces[3] eq '4' && $sortedFaces[4] eq '5') {
            $isStraight = 1;
        }
    }
    
    if ($isFlush && $isStraight) {
        my @sortedFaces = sort { $faces{$a} <=> $faces{$b} } @uniqueFaces;
        if ($sortedFaces[0] eq '10' && $sortedFaces[4] eq 'A') {
            return "Royal Straight Flush";
        }
        else {
            return "Straight Flush";
        }
    }
    elsif (scalar @uniqueFaces == 2) {
        my @faceCount = sort { $faces{$b} <=> $faces{$a} } @uniqueFaces;
        if ($faces{$faceCount[0]} == 4) {
            return "Four of a Kind";
        }
        else {
            return "Full House";
        }
    }
    elsif ($isFlush) {
        return "Flush";
    }
    elsif ($isStraight) {
        return "Straight";
    }
    elsif (scalar @uniqueFaces == 3) {
        my @faceCount = sort { $faces{$b} <=> $faces{$a} } @uniqueFaces;
        if ($faces{$faceCount[0]} == 3) {
            return "Three of a Kind";
        }
        else {
            return "Two Pair";
        }
    }
    elsif (scalar @uniqueFaces == 4) {
        return "Pair";
    }
    else {
        return "High Card";
    }
}

sub sortHands {
    my (@hands) = @_;
    
    my @sortedHands = sort {
        my $rankA = getRank(evaluateHand(@$a));
        my $rankB = getRank(evaluateHand(@$b));
        
        if ($rankA == $rankB) {
            my @sortedA = sort { $b->getFace() cmp $a->getFace() } @$a;
            my @sortedB = sort { $b->getFace() cmp $a->getFace() } @$b;
            
            for (my $i = 0; $i < 5; $i++) {
                if ($sortedA[$i]->getFace() ne $sortedB[$i]->getFace()) {
                    return $sortedB[$i]->getFace() cmp $sortedA[$i]->getFace();
                }
                elsif ($sortedA[$i]->getSuit() ne $sortedB[$i]->getSuit()) {
                    return $sortedB[$i]->getSuit() cmp $sortedA[$i]->getSuit();
                }
            }
            
            return 0;
        }
        else {
            return $rankB <=> $rankA;
        }
    } @hands;
    
    return @sortedHands;
}

sub getRank {
    my ($handType) = @_;
    
    my %ranks = (
        "Royal Straight Flush" => 10,
        "Straight Flush" => 9,
        "Four of a Kind" => 8,
        "Full House" => 7,
        "Flush" => 6,
        "Straight" => 5,
        "Three of a Kind" => 4,
        "Two Pair" => 3,
        "Pair" => 2,
        "High Card" => 1,
    );
    
    return $ranks{$handType};
}